### PR TITLE
openjdk 22

### PR DESCRIPTION
**Pull request by WangEdward**
moved to #168294

---

Help wanted: https://github.com/Homebrew/homebrew-core/pull/166788#issuecomment-2039908299

---

`openjdk` 22 and `openjdk@21` lts

fix dependency for `benerator` `carrot2` `cbmc` `flix` `glassfish` `gradle` `gssh` `jenkins` `jenkins-lts` `jmeter` `kotlin-language-server` 

---

https://blogs.oracle.com/java/post/the-arrival-of-java-22

---

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by chenrui333** [comment]
```
An exception has occurred in the compiler (22). Please file a bug against the Java compiler via the Java bug reporting page (https://bugreport.java.com/) after checking the Bug Database (https://bugs.java.com/) for duplicates. Include your program, the following diagnostic, and the parameters passed to the Java compiler in your report. Thank you.
  java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /private/tmp/openjdk-20240321-7116-vi0fz3/jdk22u-jdk-22-ga/build/macosx-aarch64-server-release/jdk/modules/jdk.jpackage/jdk/jpackage/internal/resources/jpackageapplauncher.cstemp
  	at java.base/java.nio.file.FileTreeIterator.fetchNextIfNeeded(FileTreeIterator.java:87)
  	at java.base/java.nio.file.FileTreeIterator.hasNext(FileTreeIterator.java:103)
  	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
  	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
  	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
  	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
  	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
  	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
  	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
  	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:611)
  	at java.base/jdk.internal.jrtfs.ExplodedImage.initNodes(ExplodedImage.java:259)
  	at java.base/jdk.internal.jrtfs.ExplodedImage.<init>(ExplodedImage.java:71)
  	at java.base/jdk.internal.jrtfs.SystemImage.open(SystemImage.java:79)
  	at java.base/jdk.internal.jrtfs.JrtFileSystem.<init>(JrtFileSystem.java:90)
  	at java.base/jdk.internal.jrtfs.JrtFileSystemProvider.getTheFileSystem(JrtFileSystemProvider.java:210)
  	at java.base/jdk.internal.jrtfs.JrtFileSystemProvider.getFileSystem(JrtFileSystemProvider.java:224)
  	at java.base/java.nio.file.FileSystems.getFileSystem(FileSystems.java:235)
  	at jdk.compiler/com.sun.tools.javac.file.Locations$SystemModulesLocationHandler.initSystemModules(Locations.java:1970)
  	at jdk.compiler/com.sun.tools.javac.file.Locations$SystemModulesLocationHandler.listLocationsForModules(Locations.java:1945)
  	at jdk.compiler/com.sun.tools.javac.file.Locations.listLocationsForModules(Locations.java:2197)
  	at jdk.compiler/com.sun.tools.javac.file.JavacFileManager.listLocationsForModules(JavacFileManager.java:1187)
  	at jdk.compiler/com.sun.tools.javac.code.ModuleFinder$ModuleLocationIterator.hasNext(ModuleFinder.java:139)
  	at jdk.compiler/com.sun.tools.javac.code.ModuleFinder.scanModulePath(ModuleFinder.java:295)
  	at jdk.compiler/com.sun.tools.javac.code.ModuleFinder.findAllModules(ModuleFinder.java:188)
  	at jdk.compiler/com.sun.tools.javac.comp.Modules.getUnnamedModuleCompleter(Modules.java:1458)
  	at jdk.compiler/com.sun.tools.javac.comp.Modules.setCompilationUnitModules(Modules.java:477)
  	at jdk.compiler/com.sun.tools.javac.comp.Modules.enter(Modules.java:271)
  	at jdk.compiler/com.sun.tools.javac.comp.Modules.initModules(Modules.java:237)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.initModules(JavaCompiler.java:1058)
  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:949)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:319)
  	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:178)
  	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:66)
  	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:52)
  Caused by: java.nio.file.NoSuchFileException: /private/tmp/openjdk-20240321-7116-vi0fz3/jdk22u-jdk-22-ga/build/macosx-aarch64-server-release/jdk/modules/jdk.jpackage/jdk/jpackage/internal/resources/jpackageapplauncher.cstemp
  	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
  	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
  	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
  	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:56)
  	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
  	at java.base/java.nio.file.Files.readAttributes(Files.java:1858)
  	at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:220)
  	at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:277)
  	at java.base/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:374)
  	at java.base/java.nio.file.FileTreeIterator.fetchNextIfNeeded(FileTreeIterator.java:83)
  	... 33 more
  printing javac parameters to: /private/tmp/openjdk-20240321-7116-vi0fz3/jdk22u-jdk-22-ga/make/javac.20240321_095730.args
  make[3]: *** [/private/tmp/openjdk-20240321-7116-vi0fz3/jdk22u-jdk-22-ga/build/macosx-aarch64-server-release/buildtools/create_symbols_javac/_the.COMPILE_CREATE_SYMBOLS_batch] Error 4
  make[2]: *** [jdk.compiler-gendata] Error 2
  make[2]: *** Waiting for unfinished jobs....
  
  ERROR: Build failed for target 'images' in configuration 'macosx-aarch64-server-release' (exit code 2)
```

File: Formula/o/openjdk.rb [hunk 1]
```diff
@@ -55,22 +55,22 @@ class Openjdk < Formula
   resource "boot-jdk" do
     on_macos do
       on_arm do
-        url "https://download.java.net/java/GA/jdk21.0.1/415e3f918a1f4062a0074a2794853d0d/12/GPL/openjdk-21.0.1_macos-aarch64_bin.tar.gz"
```
Comment (SMillerDev) [inline Formula/o/openjdk.rb:58]: Let's keep the boot JDKs on the LTS version

**Review by SMillerDev — COMMENTED** [review]
Are all the patches still needed?

**Comment by ryenus** [comment]
Regarding the test error with **benerator**:

> `brew test --retry --verbose benerator` failed on macOS Sonoma (14) on Apple Silicon!
> ==> Testing benerator
> ...
> Exception in thread "main" java.lang.NoSuchMethodError: 'void sun.misc.Unsafe.ensureClassInitialized(java.lang.Class)'
> 	at com.oracle.truffle.api.library.LibraryFactory.ensureLibraryInitialized(LibraryFactory.java:384)

This appears to be related to the removal of `sun.misc.Unsafe.ensureClassInitialized`, as mentioned in the [section "Removed Features and Options" of JDK 22 Release Notes](https://www.oracle.com/java/technologies/javase/22all-relnotes.html#Removed)
> sun.misc.Unsafe.shouldBeInitialized and ensureClassInitialized Are Removed ([JDK-8316160](https://bugs.openjdk.org/browse/JDK-8316160))

Meanwhile according to https://github.com/rapiddweller/rapiddweller-benerator-ce/blob/development/pom.xml#L123-L124, benerator should **NOT** depends on **`openjdk`**, but **`openjdk@11`**:

> ```xml
> <maven.compiler.source>11</maven.compiler.source>
> <maven.compiler.target>11</maven.compiler.target>
> ```

**Comment by WangEdward** [comment]
## Build Failures 
`openjdk` on **Linux**
```
==> brew install ./openjdk--22.x86_64_linux.bottle.tar.gz
==> brew linkage --test openjdk
==> FAILED
Full linkage --test openjdk output
  Missing libraries:
    libjava.so
```

## Dependency Failures
- Fixed: `benerator` `carrot2` `cbmc?` `flix` `glassfish` `gradle` `gssh` `jenkins` `jenkins-lts` `jmeter` `kotlin-language-server` 
- Fomulea after `mercury` need more time to be tested https://github.com/Homebrew/homebrew-core/actions/runs/8553455945/job/23438377005
### Not related
- `csound` => `libsndfile` => `flac` => `homebrew_curl` : `Download failed: Homebrew-installed 'curl' is not installed for https://...`
- `joern` => `freetds`  => `homebrew_curl`

File: Formula/j/jenkins-lts.rb [hunk 2]
```diff
@@ -14,10 +14,10 @@ class JenkinsLts < Formula
...
 
-  depends_on "openjdk"
+  depends_on "openjdk@21"
 
   def install
-    system "#{Formula["openjdk"].opt_bin}/jar", "xvf", "jenkins.war"
+    system "#{Formula["openjdk@21"].opt_bin}/jar", "xvf", "jenkins.war"
     libexec.install "jenkins.war", "WEB-INF/lib/cli-#{version}.jar"
     bin.write_jar_script libexec/"jenkins.war", "jenkins-lts"
     bin.write_jar_script libexec/"cli-#{version}.jar", "jenkins-lts-cli"
```
Comment (ZhongRuoyu) [inline Formula/j/jenkins-lts.rb:23]:
```suggestion
    bin.write_jar_script libexec/"jenkins.war", "jenkins-lts", java_version: "21"
    bin.write_jar_script libexec/"cli-#{version}.jar", "jenkins-lts-cli", java_version: "21"
```

File: Formula/b/benerator.rb
Comment (ZhongRuoyu) [inline Formula/b/benerator.rb:1]: `s/overridable_java_home_env/overridable_java_home_env("21")/`

File: Formula/f/flix.rb
Comment (ZhongRuoyu) [inline Formula/f/flix.rb:1]: `s/write_jar_script(.*)/write_jar_script\1, java_version: "21"/g`

(I.e., append `java_version: "21"` to calls to `write_jar_script`.)

File: Formula/g/glassfish.rb
Comment (ZhongRuoyu) [inline Formula/g/glassfish.rb:1]: `s/overridable_java_home_env/overridable_java_home_env("21")/`

File: Formula/g/gradle.rb
Comment (ZhongRuoyu) [inline Formula/g/gradle.rb:1]: `s/overridable_java_home_env/overridable_java_home_env("21")/`

File: Formula/g/gssh.rb
Comment (ZhongRuoyu) [inline Formula/g/gssh.rb:1]: `s/write_jar_script(.*)/write_jar_script\1, java_version: "21"/g`

File: Formula/j/jenkins.rb
Comment (ZhongRuoyu) [inline Formula/j/jenkins.rb:1]: `s/write_jar_script(.*)/write_jar_script\1, java_version: "21"/g`

**Review by ZhongRuoyu — COMMENTED** [review]
Nice work @WangEdward!

Can you please also make sure the commits follow our "one formula per commit; one commit per formula" rule and are formatted accordingly (e.g. `` benerator: switch to `openjdk@11` ``)? You can refer to https://docs.brew.sh/Formula-Cookbook#commit for details.

**Comment by ZhongRuoyu** [comment]
Superseded by #168294

**Pull request closed without merging**